### ISSUE TITLE
Implement a more consistent behavior for evaluation errors

### ISSUE THREAD

**Issue by Pankraty** [ClosedXML/ClosedXML#826]
Inspired by #824, #585 

## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [ ] Bug
- [X] Feature

**Version of ClosedXML**

0.93-beta2

**What is the current behavior?**

Dealing with errors occurring on formula evaluation is quite hard and counter-intuitive. Say, there are functions `HLOOKUP` and `VLOOKUP` that return `#N/A` when the value is not found. In ClosedXML they throw the exception making impossible to evaluate formulas that include such functions or refer to them (i.e. `=IF(ISNA(HLOOKUP(...), "Not found", ...)`).

**What is the expected behavior or new feature?**

1. Reading the value of the cell containing #N/A, #REF or other types of errors should not throw exceptions.
2. If the evaluation of the cell formula produces an error the `Value` of the cell should return the code of the error (enum containing codes for #NA, #REF, #DIV/0, etc.) and the `DataType` of the cell should be `Error` (a new value in the existing enum).
3. Putting a string representation of the error code (like '#N/A') should set the cell type to `Error` too (this is disputable, but that's how it works in Excel).
4. A text representation of the error code produced by formula (e.g. `="#N" & "/A"`) must not be treated as an error. 
5. Cells having errors should be possible to be saved to file and to be read from the file.

The same approach should be applied regarding all types of errors.

**Did this work in previous versions of our tool?  Which versions?**

Not that I knew

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

Without a code sample, it is unlikely that your issue will get attention. Don't be lazy. Do the effort and assist the developers to reproduce your problem. Code samples should be [minimal complete and verifiable](https://stackoverflow.com/help/mcve). Sample spreadsheets should be attached whenever applicable. Remove sensitive information.

**Code to reproduce problem:**
```c#
public void Main()
{
    //This code should not throw exceptions
    var ws = wb.Worksheets.Add("Sheet1");

    ws.Cell("A1").FormulaA1 = "=\"#N\" & \"/A\"";
    ws.Cell("A2").Value = "#N/A";
    ws.Cell("A3").FormulaA1 = "=HLOOKUP(0,C1:D3,1,FALSE)";

    ws.Range("B1:B3").FormulaR1C1 = "=ISNA(RC[-1])";

    Assert.AreEqual(XLDataType.Text, ws.Cell("A1").DataType);
    Assert.AreEqual("#N/A", ws.Cell("A1").Value);
    Assert.IsFalse((bool)ws.Cell("B1").Value);

    Assert.AreEqual(XLError.NA, ws.Cell("A2").Value);  // A new enum
    Assert.AreEqual(XLDataType.Error, ws.Cell("A2").DataType); // A new DataType
    Assert.IsTrue((bool)ws.Cell("B2").Value);

    Assert.AreEqual(XLError.NA, ws.Cell("A3").Value);  // A new enum
    Assert.AreEqual(XLDataType.Error, ws.Cell("A3").DataType); // A new DataType
    Assert.IsTrue((bool)ws.Cell("B3").Value);
}
```
- [x] I attached a sample spreadsheet.  (You can drag files on to this issue)
[NA.xlsx](https://github.com/ClosedXML/ClosedXML/files/1963657/NA.xlsx)

**Comment by igitur** [comment]
Yes, even when I implemented the exceptions, I thought about this approach. But before my implementation, formulas would just throw arbitrary .NET exceptions, so at least there was some improvement.

I still wanted to check how Excel manages different errors in one formula. e.g. suppose there is some complex, nested formula, where some part returns N/A and some part returns REF. Does the ultimate return value of the cell depend on the first error that is encountered or is there some preferred order to errors and that the 'most important' error is returned.

**Comment by igitur** [comment]
Went through your code and I agree with most part. I just dislike the abbreviated `NA`. I'd prefer the full name for the enum. Similar to the current exceptions.

I think the current throwing of exceptions could be useful for some people who want to ensure that their worksbooks are error free. I'm thinking we should add a `LoadOptions` struct which would include an option for whether the user wants `XLError` values or exceptions to be thrown.

**Comment by Pankraty** [comment]
> I just dislike the abbreviated NA. I'd prefer the full name for the enum. 

No problem. This issue was created to gather requirements and to discuss the implementation details.

> I'm thinking we should add a LoadOptions struct which would include an option for whether the user wants XLError values or exceptions to be thrown.

How about doing this not in `LoadOptions` but in `XLWorkbook.RecalculateAllFormulas()`? It has not been released in non-beta version so we are free to change its signature (e.g. add an argument `EvaluationOptions`)

**Comment by igitur** [comment]
> How about doing this not in LoadOptions but in XLWorkbook.RecalculateAllFormulas()? It has not been released in non-beta version so we are free to change its signature (e.g. add an argument EvaluationOptions)

Agreed.

Also, we already have this: `internal enum ExpressionErrorType`. We could reuse that.

**Comment by jahav** [comment]
The issue has been implemented through PR #1934, #1936, #1935 and #1404.

It corresponds to the feature described in the original post, except for setting the value though string value. The heuristic "#N/A" -> guess `#N/A` error is an Excel interface for humans and only in default Excel configuration (e.g. go to cell, change number format from *General* to *Text*, enter *#N/A* to the cell, switch the format back to *General* -> cell has a string value *#N/A*).

ClosedXML has an interface for code and it should have a clear semantic. That has been mostly done in #1894 and the value should be set like this:

```csharp
ws.Cell(1,1).Value = XLError.IncompatibleValue; // Cell now contains an error
ws.Cell(2,1).Value = "#N/A"; // Cell contains a text '#N/A'
```
